### PR TITLE
Introduce `.spec.schedulerName` in `Shoot` API

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area usability
/kind enhancement

**What this PR does / why we need it**:
This PR adds a new field `.spec.schedulerName` for `Shoot`s. Similar to the `schedulerName` field in `Pod`s, the configuration allows to address specific shoot schedulers in a landscape.

The new `default-scheduler` name is reserved for Gardener's "built-in" scheduler. All `Shoot`s not defining this field, are defaulted to this configuration.

**Special notes for your reviewer**:
/cc @rfranzke

**Release note**:

```feature operator
`Shoot`s allow to optionally configure a specific scheduler via `.spec.schedulerName`. The `default-scheduler` is used in case non is configured. Please note, that `Shoot`s will remain `Pending` in case a scheduler name is configured but an adequate scheduler is not available in the landscape.
```

### CODE REVIEW

File: docs/concepts/scheduler.md [hunk 1]
```diff
@@ -87,6 +87,10 @@ The region does not matter, i.e., `testing` shoots may also be scheduled on a se
 The `shoots/binding` subresource is used to bind a `Shoot` to a `Seed`. On creation of a shoot cluster/s, the scheduler updates the binding automatically if an appropriate seed cluster is available.
 Only an operator with the necessary RBAC can update this binding manually. This can be done by changing the `.spec.seedName` of the shoot. However, if a different seed is already assigned to the shoot, this will trigger a control-plane migration. For required steps, please see [Triggering the Migration](../operations/control_plane_migration.md#triggering-the-migration).
 
+## `spec.schedulerName` Field in the `Shoot` Specification
+Similar to the `spec.defaultScheduler` field in `Pod`s, the `Shoot` specification has an optional `.spec.schedulerName` field. If this field is set on creation, only the scheduler which relates to the configured name is responsible for scheduling the shoot.
+The `default-scheduler` name is reserved for the default scheduler of Gardener.
+
```
Comment (ary1992) [inline docs/concepts/scheduler.md:95]:
```suggestion
Shoot will remain in `Pending` state if the mentioned scheduler is not present in the landscape.

```

File: pkg/apis/core/validation/shoot.go [hunk 2]
```diff
@@ -327,6 +328,13 @@ func ValidateShootSpecUpdate(newSpec, oldSpec *core.ShootSpec, newObjectMeta met
 
 	allErrs = append(allErrs, validateNetworkingUpdate(newSpec.Networking, oldSpec.Networking, fldPath.Child("networking"))...)
 
+	if !reflect.DeepEqual(oldSpec.SchedulerName, newSpec.SchedulerName) {
+		// only allow to set an empty scheduler name to the default scheduler
+		if oldSpec.SchedulerName != nil || pointer.StringDeref(newSpec.SchedulerName, "") != v1beta1constants.DefaultSchedulerName {
```
Comment (ary1992) [inline pkg/apis/core/validation/shoot.go:333]: With defaulting in place, will the `oldSpec.SchedulerName` be ever set to nil ?
Moreover, why we are restricting users to only set `default scheduler`?

File: pkg/apis/core/validation/shoot_test.go [hunk 3]
```diff
@@ -4527,6 +4527,31 @@ var _ = Describe("Shoot Validation Tests", func() {
...
+				oldShoot.Spec.SchedulerName = pointer.String("foo-scheduler")
+
+				errorList := ValidateShootSpecUpdate(&shoot.Spec, &oldShoot.Spec, metav1.ObjectMeta{}, field.NewPath("spec"))
+
+				Expect(errorList).To(ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{
+					"Type":  Equal(field.ErrorTypeInvalid),
+					"Field": Equal("spec.schedulerName"),
+				}))))
+			})
+		})
```
Comment (ary1992) [inline pkg/apis/core/validation/shoot_test.go:4567]: Maybe add a test for the case where `shoot.Spec.SchedulerName` is not nil and scheduler is changed from value `a` to `b`

File: pkg/scheduler/controller/shoot/add.go [hunk 4]
```diff
@@ -54,7 +56,8 @@ func (r *Reconciler) AddToManager(mgr manager.Manager) error {
 func (r *Reconciler) ShootPredicate() predicate.Predicate {
 	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
 		if shoot, ok := obj.(*gardencorev1beta1.Shoot); ok {
-			return shoot.Spec.SeedName == nil
+			return shoot.Spec.SeedName == nil &&
+				pointer.StringDeref(shoot.Spec.SchedulerName, v1beta1constants.DefaultSchedulerName) == v1beta1constants.DefaultSchedulerName
```
Comment (ary1992) [inline pkg/scheduler/controller/shoot/add.go:61]: Adapt the doc string also.

File: pkg/scheduler/controller/shoot/add_test.go [hunk 5]
```diff
@@ -83,5 +83,33 @@ var _ = Describe("Add", func() {
...
 			})
 		})
+
+		Context("shoot defines schedulerName", func() {
+			Context("default-scheduler", func() {
+				BeforeEach(func() {
+					shoot.Spec.SchedulerName = pointer.String("default-scheduler")
+				})
+
+				It("should be false", func() {
```
Comment (ary1992) [inline pkg/scheduler/controller/shoot/add_test.go]:
```suggestion
                                It("should be true", func() {
```

**Review by ary1992 — COMMENTED** [review]
Thanks for the PR.
Overall looks good ,left few comments.

File: pkg/apis/core/validation/shoot.go (hunk 2 above)
Comment (timuthy) [reply pkg/apis/core/validation/shoot.go:333]: The field of existing shoots is `nil`. The idea was to set the `schedulerName` field for these.

File: docs/concepts/scheduler.md [hunk 6]
```diff
@@ -87,6 +87,10 @@ The region does not matter, i.e., `testing` shoots may also be scheduled on a se
 The `shoots/binding` subresource is used to bind a `Shoot` to a `Seed`. On creation of a shoot cluster/s, the scheduler updates the binding automatically if an appropriate seed cluster is available.
 Only an operator with the necessary RBAC can update this binding manually. This can be done by changing the `.spec.seedName` of the shoot. However, if a different seed is already assigned to the shoot, this will trigger a control-plane migration. For required steps, please see [Triggering the Migration](../operations/control_plane_migration.md#triggering-the-migration).
 
+## `spec.schedulerName` Field in the `Shoot` Specification
+Similar to the `spec.defaultScheduler` field in `Pod`s, the `Shoot` specification has an optional `.spec.schedulerName` field. If this field is set on creation, only the scheduler which relates to the configured name is responsible for scheduling the shoot.
```
Comment (rfranzke) [inline docs/concepts/scheduler.md]:
```suggestion

Similar to the `spec.schedulerName` field in `Pod`s, the `Shoot` specification has an optional `.spec.schedulerName` field. If this field is set on creation, only the scheduler which relates to the configured name is responsible for scheduling the shoot.
```

File: docs/concepts/scheduler.md [hunk 7]
```diff
@@ -87,6 +87,10 @@ The region does not matter, i.e., `testing` shoots may also be scheduled on a se
...
+The `default-scheduler` name is reserved for the default scheduler of Gardener.
+
 ## `spec.seedName` Field in the `Shoot` Specification
```
Comment (rfranzke) [inline docs/concepts/scheduler.md:96]:
```suggestion
## `spec.seedName` Field in the `Shoot` Specification

```

File: docs/concepts/scheduler.md [hunk 8]
```diff
@@ -116,4 +120,4 @@ In case the scheduler fails to find a suitable seed, the operation is being retr
 
 ## Current Limitation / Future Plans
 
-- Azure unfortunately has a geographically non-hierarchical naming pattern and does not start with the continent. This is the reason why we will exchange the implementation of the `MinimalDistance` strategy with a more suitable one in the future.
\ No newline at end of file
+- Azure unfortunately has a geographically non-hierarchical naming pattern and does not start with the continent. This is the reason why we will exchange the implementation of the `MinimalDistance` strategy with a more suitable one in the future.
```
Comment (rfranzke) [inline docs/concepts/scheduler.md:126]: 👀

File: example/provider-local/shoot-workerless.yaml [hunk 9]
```diff
@@ -4,7 +4,8 @@ metadata:
   name: local-wl
   namespace: garden-local
 spec:
-  seedName: local
+  #schedulerName: foo
+  #seedName: local
```
Comment (rfranzke) [inline example/provider-local/shoot-workerless.yaml]:
```suggestion
```

File: pkg/scheduler/controller/shoot/reconciler.go [hunk 10]
```diff
@@ -70,6 +72,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 		return reconcile.Result{}, nil
 	}
 
+	if pointer.StringDeref(shoot.Spec.SchedulerName, v1beta1constants.DefaultSchedulerName) != v1beta1constants.DefaultSchedulerName {
+		log.Info("Ignoring shoot because it requests a different scheduler")
+		return reconcile.Result{}, nil
+	}
```
Comment (rfranzke) [inline pkg/scheduler/controller/shoot/reconciler.go]: (almost) dead code ☠️

File: docs/concepts/scheduler.md (hunk 8 above)
Comment (timuthy) [reply docs/concepts/scheduler.md:126]: No clue where this diff comes from ...

File: example/provider-local/shoot.yaml [hunk 11]
```diff
@@ -7,7 +7,6 @@ metadata:
     shoot.gardener.cloud/infrastructure-cleanup-wait-period-seconds: "0"
     shoot.gardener.cloud/cloud-config-execution-max-delay-seconds: "0"
 spec:
-  seedName: local
```
Comment (rfranzke) [inline example/provider-local/shoot.yaml:10]: Also update https://github.com/gardener/gardener/blob/master/example/provider-local/managedseeds/shoot-managedseed.yaml (maybe update Kubernetes version as well)

**Review by rfranzke — COMMENTED** [review]
Nice, thanks!
/approve

File: pkg/apis/core/v1beta1/defaults_test.go [hunk 12]
```diff
@@ -313,6 +313,22 @@ var _ = Describe("Defaults", func() {
...
 
+		It("should default schedulerName", func() {
+			obj.Spec.SchedulerName = nil
+
+			SetObjectDefaults_Shoot(obj)
+
+			Expect(obj.Spec.SchedulerName).To(PointTo(Equal("default-scheduler")))
+		})
+
+		It("should default schedulerName", func() {
```
Comment (oliver-goetz) [inline pkg/apis/core/v1beta1/defaults_test.go]:
```suggestion
		It("should set the custom schedulerName", func() {
```

**Review by oliver-goetz — COMMENTED** [review]
Nice PR, I just found one nit 🚀 
and I have a question. Why is the `schedulerName` immutable?

**Comment by timuthy** [comment]
@oliver-goetz thanks for the review
> Why is the schedulerName immutable?

Because once set, only the responsible scheduler should act on this event. If the name is changed again right after creation, then two schedulers might interfere with each other.

**Pull request merged**